pubsub/envelope.go

```diff
@@ -110,6 +115,40 @@ func NewCloudEventsEnvelope(id, source, eventType, subject string, topic string,
 	}
 }
 
+// HasExpired determines if the current cloud event has expired.
+func (cloudEvent *CloudEventsEnvelope) HasExpired() bool {
+	if cloudEvent.Expiration != "" {
+		expiration, err := time.Parse(time.RFC3339, cloudEvent.Expiration)
+		if err != nil {
+			return false
+		}
+
+		return expiration.UTC().Before(time.Now().UTC())
+	}
+
+	return false
+}
+
+// ApplyMetadata will process metadata to modify the cloud event based on the component's feature set.
+func (cloudEvent *CloudEventsEnvelope) ApplyMetadata(componentFeatures []Feature, metadata map[string]string) {
+	ttl, hasTTL, _ := contrib_metadata.TryGetTTL(metadata)
+	if hasTTL && !FeatureMessageTTL.IsPresent(componentFeatures) {
+		// Dapr only handles Message TTL if component does not.
+		now := time.Now().UTC()
+		expiration := now.Add(ttl)
+		// The maximum duration (maxInt64) is not enough to overflow max time, for now.
+		// As of the time this code was written (2020 Dec 28th),
+		// the maximum time of now() adding maxInt64 is ~ "2313-04-09T23:30:26Z".
+		// Max time in golang is currently 292277024627-12-06T15:30:07.999999999Z.
+		// So, we have some time before the overflow below happens.
+		if (ttl.Seconds() > 0) && expiration.Before(now) {
+			// Overflow
+			expiration = time.Unix(1<<63-62135596801, 999999999).UTC()
+		}
```

Comment:
Would it be better to throw an error of some kind rather than silently set this to max time? If a users code is overflowing, that's probably an indicator of a bug in the user code. Shouldn't we expose this error sooner, rather than make the user investigate why their messages are never expiring?

Also, is there a constant available for these magic numbers?

Reply:
After carefully re-reading this comment, is the assertion that this is code we will never actually hit? If so, shouldn't we panic here?

Reply:
This code will "never" run. I actually thought of leaving the comment explaining why the overflow will unlikely happen and remove this if clause.